### PR TITLE
Only derive Clone on builders when necessary

### LINKED ISSUE #97: Owned builders should not #[derive(Clone)]
Hi,

my struct has a non-clonable field, thus i want to disable the derived impl of *Clone* trait for the generated builder.

### PR DESCRIPTION
**Pull request by TedDriggs**
Builders using the owned pattern for all fields don't need Clone derived for them,
and therefore can be used with generics that don't support cloning.

Fixes #97, and obviates #111 

@azriel91, please review this.

### CODE REVIEW

**Comment by azriel91** [comment]
I'll get to this in the upcoming Tuesday, going to be temporarily inactive till then

File: derive_builder_core/src/builder.rs [hunk 1]
```diff
@@ -84,6 +89,24 @@ impl<'a> ToTokens for Builder<'a> {
...
+
+            // Create the comma-separated set of derived traits for the builder
+            let derived_traits = {
+                let default_trait: Ident = parse_quote!(Default);
+                let clone_trait: Ident = parse_quote!(Clone);
+
+                let mut traits: Punctuated<&Ident, Token![,]> = Default::default();
+                traits.push(&default_trait);
+
+                if self.must_derive_clone {
```
Comment (azriel91) [inline derive_builder_core/src/builder.rs]: Should this also check if the builder pattern is non-owned?
The tests use the `default_builder!` macro, which has `must_derive_clone = true`, but `Options#as_builder` only checks if any fields require clone. I've tried tracing the code but haven't found the answer

File: derive_builder/src/options/darling_opts.rs [hunk 2]
```diff
@@ -319,6 +319,16 @@ impl Options {
             .fields
     }
 
+    /// A builder requires `Clone` to be derived if any of its fields use the mutable or immutable pattern.
+    ///
+    /// This can technically be expressed two ways:
```
Comment (azriel91) [inline derive_builder/src/options/darling_opts.rs]: Feels weird to explain how to express that `Clone` is required, and it should be the non-owned patterns that cause `Clone` to be required. I think the first sentence by itself is enough

File: derive_builder_core/src/builder.rs (hunk 1 above)
Comment (TedDriggs) [reply derive_builder_core/src/builder.rs]: This shouldn't, but I was wondering if the function in `darling_opts` should.

File: derive_builder/src/options/darling_opts.rs (hunk 2 above)
Comment (TedDriggs) [reply derive_builder/src/options/darling_opts.rs]: You're right; I'll remove the second part of this comment.

File: derive_builder/src/options/darling_opts.rs [hunk 3]
```diff
@@ -319,6 +319,16 @@ impl Options {
...
+    ///
+    /// 1. The caller chooses the `owned` pattern at the struct level
+    /// 1. The caller chooses the `owned` pattern on every field
+    pub fn requires_clone(&self) -> bool {
+        self.fields().any(|f| f.pattern().requires_clone())
```
Comment (TedDriggs) [inline derive_builder/src/options/darling_opts.rs]: @azriel91 if there's a place where we should check the struct's pattern, it'd be here. However, I don't think we strictly _need_ to; unless I'm missing something, it only matters what the individual fields are doing: A mutable zero-field builder shouldn't need `Clone` to operate.

Comment (azriel91) [reply derive_builder/src/options/darling_opts.rs]: True that (though one would wonder why you'd have a zero field builder :stuck_out_tongue:)

Comment (TedDriggs) [reply derive_builder/src/options/darling_opts.rs]: Wait, I forgot that `build` takes a `&self` if the struct-level one isn't marked `owned`. I've updated the logic accordingly.

**Pull request merged**